packages/loader/container-loader/src/container.ts

```diff
@@ -2011,7 +2036,22 @@ export class Container
 
 		deltaManager.on("connect", (details: IConnectionDetailsInternal, _opsBehind?: number) => {
 			assert(this.connectionMode === details.mode, 0x4b7 /* mismatch */);
-			this.connectionStateHandler.receivedConnectEvent(details);
+
+			// Delay raising events until setLoaded()
+			// Here are some of the reasons why this design is chosen:
+			// 1. Various processes track speed of connection. But we are not processing ops or signal while container is loading,
+			//    and thus we can't move forward across connection modes. This results in telemetry errors (like NoJoinOp) that
+			//    have nothing to do with connection flow itself
+			// 2. This also makes it hard to reason about recovery (like reconnection) in case we might have lost JoinSignal. Reconnecting
+			//    in loading phase is useless (get back to same state), but at the same time not doing it may result in broken connection
+			//    without recovery (after we loaded).
+			// 3. We expose non-consistent view. ContainerRuntime may start loading in non-connected state, but end in connected, with
+			//    no events telling about it (until we loaded). Most of the code relies on a fact that state changes when events fire.
+			// This will not delay any processes (as observed by the user). I.e. once container moves to loaded phase,
+			// we immediately would transition across all phases, if we have proper signals / ops ready.
+			if (this.loaded) {
+				this.connectionStateHandler.receivedConnectEvent(details);
+			}
```

Comment:
We will deliver that same event in setLoaded(), after container is loaded, and only if it's still connected.

That said, nomenclature is very confusing. "connect" is internal event (of sorts), it maps to ConnectionState.CatchingUp that we actually do not expose (mostly) to the users. We expose only "connected" event which maps to ConnectionState.Connected, and that could happen (with this PR) only after container is loaded (as we do not process ops and signals through that process).

That all said, I think I have an issue with current code. Or/and some claims I make might not be entirely true and there is a chance to simplify. I'm making a statement (comments in setLoaded() and calling this.setContextConnectedState() vs. more inclusive this.propagateConnectionState()) that we can't be there in "connected" state. But this does not look like true (and not asserted). We call this.handleDeltaConnectionArg() a moment earlier and that will allow ops and signals to go through, which (if we are lucky) could immidiatly transition us to "connected" event (in "loading", not "loaded" state).

Let me take another look at it...

Reply:
Assert was there, but race was there too. Made the following changes:
- Added assert in handleDeltaConnectionArg() that it can be called only in this.loaded state.
- Change the order of this.setLoaded() and this.handleDeltaConnectionArg() to satisfy requirement above.

Now it's guaranteed that we will unpause op & signal processing only after container moves to loaded state, and thus it's guaranteed that "connected" transition can happen only in loaded state.

That said, as mentioned before, this code ("connect" handler) can run in loading state, and setLoaded() will deliver event after we are loaded.